utils/src/main/java/software/amazon/awssdk/utils/async/SimplePublisher.java

```diff
@@ -0,0 +1,497 @@
+
+    /**
+     * The queue of events to be processed, in the order they should be processed.
+     *
+     * <p>All logic within this publisher is represented using events in this queue. This ensures proper ordering of events
+     * processing and simplified reasoning about thread safety.
+     */
+    private final Queue<QueueEntry<T>> eventQueue = new ConcurrentLinkedQueue<>();
+
+    /**
+     * When processing the {@link #eventQueue}, these are the entries that should be skipped (and failed). This is used to
+     * safely drain the queue when there are urgent events needing processing, like a {@link Subscription#cancel()}.
+     */
+    private final Set<QueueEntry.Type> entryTypesToFail = new CopyOnWriteArraySet<>();
+
+    /**
+     * Whether the {@link #eventQueue} is currently being processed. Only one thread may read events from the queue at a time.
+     */
+    private final AtomicBoolean processingQueue = new AtomicBoolean(false);
+
+    /**
+     * An exception that should be raised to any failed {@link #write(Object)}, {@link #complete()} or {@link #error(Throwable)}
+     * operations. This is used to stop accepting messages after the downstream subscription is cancelled or after the
+     * caller sends a {@code complete()} or {@code #error()}.
+     *
+     * <p>This is a supplier to avoid the cost of creating an exception in the successful code path.
+     */
+    private final AtomicReference<Supplier<RuntimeException>> rejectException = new AtomicReference<>();
+
+    /**
+     * The subscriber provided via {@link #subscribe(Subscriber)}. This publisher only supports a single subscriber.
+     */
+    private Subscriber<? super T> subscriber;
+
+    /**
+     * Write a message to this publisher.
+     *
+     * <p>Messages written to this publisher will eventually be sent to a downstream subscriber, in the order they were
+     * written. When the message is sent to the subscriber, the returned future will be completed successfully.
+     *
+     * <p>This method may be invoked concurrently when the order of messages is not important.
+     *
+     * <p>In the time between when this method is invoked and the returned future is not completed, this publisher stores the
+     * request message in memory. Callers are recommended to limit the number of writes in progress at a time to bound the
+     * amount of memory used by this publisher.
+     *
+     * <p>The returned future will be completed exceptionally if the downstream subscriber cancels the subscription, or
+     * if the write call was performed after a {@link #complete()} or {@link #error(Throwable)} call.
+     *
+     * @param value The message to send. Must not be null.
+     * @return A future that is completed when the message is sent to the subscriber.
+     */
+    public CompletableFuture<Void> write(T value) {
+        log.trace(() -> "Received write() with " + value);
+
+        OnNextQueueEntry<T> entry = new OnNextQueueEntry<>(value);
```

Comment:
Query : Just curious , since we are creating New instances then previous implementation does this qualifies to a quick performance/memory test ?

Reply:
Creating an instance in the JVM is pretty quick (mostly equivalent to work done in the constructor), and short-lived instances are very efficiently GC'd.

Creating this instance is slower than not creating one, but it's a bit faster than copying the bytes out of the buffer will be, a fair bit faster than the volatile reads/writes we're doing, and considerably faster than the network activity that all of this relies on.

That said, this is all supposition. The only way to know for sure would be to do profiling, and the output might be surprising. I intend to bake these changes in the canaries for a little while to make sure that it's not slower in any noticeable way.

I expect that this implementation might be slightly slower than the previous implementation, but I think it will make us more productive because it puts all of the confusing publisher stuff into a separate class that we (hopefully) don't have to think about again.